### ISSUE TITLE
Distinction between error types

### ISSUE THREAD

**Issue by chamook** [Zaid-Ajaj/Npgsql.FSharp#70]
Currently there are a few points that return a `System.Exception` as the failure type for a few different errors, and AFAICT only really provide a different message to distinguish between them. This led to me writing this somewhat hacky piece of code today to distinguish between errors internally:

```fsharp
let private readDatabaseError (error: exn) =
    let notFoundMessage =
        "Expected at least one row to be returned from the result set"
    match error with
    | e when e.Message.StartsWith notFoundMessage -> NoRecord
    | _ -> Unknown error
```

I think it could be an improvement to raise different exceptions depending on the cause of the failure, but that would also be a breaking change so I'm not sure if it would be more hassle than it's worth. FWIW I would be happy to put some time into working on this if you think it should be changed.

What do you think?

**Comment by Zaid-Ajaj** [comment]
Hello there @chamook, 

I think it is a great idea to add a specialized exception type when raising the error about one expected row to be returned. If you could a PR to implement a new exception type and raise it inside `Sql.executeRow` and `Sql.executeRowAsync`! 

Other types of exceptions are actually coming from Npgsql so you can actually cast the exception to `PostgresException` and inspect it for more information such as whether it was a syntax error, a query timeout or asking for tables that don't exist.

**Comment by Zaid-Ajaj** [comment]
Published v3.9 :rocket: and now it includes the custom error types. Thanks a lot for the contribution 🙏
